common/ledger/snapshot/file.go

```diff
@@ -0,0 +1,200 @@
+/*
+Copyright IBM Corp. All Rights Reserved.
+
+SPDX-License-Identifier: Apache-2.0
+*/
+
+package snapshot
+
+import (
+	"bufio"
+	"encoding/binary"
+	"fmt"
+	"hash"
+	"io"
+	"os"
+
+	"github.com/golang/protobuf/proto"
+	"github.com/pkg/errors"
+)
+
+// FileCreator creates a new file for ledger snapshot. This is expected to be used by various
+// components of ledger, such as blockstorage and statedb for exporting the relevent snapshot data
+type FileCreator struct {
+	file              *os.File
+	dataformat        byte
+	hasher            hash.Hash
+	bufWriter         *bufio.Writer
+	multiWriter       io.Writer
+	varintReusableBuf []byte
+}
+
+// CreateFile creates a new file for exporting the ledger snapshot data
+// This function returns an error if the file already exists. The `dataformat` is the first byte
+// written to the file. The hasher is used for computing the hash of the data stream
+func CreateFile(filePath string, dataformat byte, hasher hash.Hash) (*FileCreator, error) {
+	file, err := os.OpenFile(filePath, os.O_RDWR|os.O_CREATE|os.O_EXCL, 0444)
+	if err != nil {
+		return nil, errors.Wrapf(err, "error while creating the snapshot file: %s", filePath)
+	}
+	bufWriter := bufio.NewWriter(file)
+	multiWriter := io.MultiWriter(bufWriter, hasher)
+	if _, err := multiWriter.Write([]byte{dataformat}); err != nil {
+		file.Close()
+		return nil, errors.Wrapf(err, "error while writing data format to the snapshot file: %s", filePath)
+	}
+	return &FileCreator{
+		file:              file,
+		bufWriter:         bufWriter,
+		multiWriter:       multiWriter,
+		hasher:            hasher,
+		varintReusableBuf: make([]byte, binary.MaxVarintLen64),
+	}, nil
+}
+
+// EncodeString encodes and appends the string to the data stream
+func (c *FileCreator) EncodeString(str string) error {
```

Comment:
```suggestion
func (c *FileWriter) WriteString(str string) error {
```
`Writer.WriteXXX` seems like a correct usage as per `bufio` and other golang io packages. 

By using `WriteString()`, we aren't really hiding some important info from the consumer. In fact, `encoding/decoding` and `length` have now become internal detail, and consumers need not be aware of it. 

The abstraction that we expose to the consumer is reading or writing the following from/to a file -- `string`, `proto`, `[]byte`, `uint64`

1. `EndoceString()` -> `WriteString()`
1. `EncodeProtoMessage()` -> `WriteProtoMessage()`
2. `EncodeBytes()` -> `WriteBytes()`
3. `EncodeUVarint()` -> `WriteUint64()`

Reply:
`bufio` and other golang package that you may have in mind do not modify the bytes supplied and hence they are pure writers. This usages resembles more with `proto` (https://github.com/golang/protobuf/blob/master/proto/buffer.go#L140) package

Reply:
[bufio.WriteRune()](https://github.com/golang/go/blob/master/src/bufio/bufio.go#L672) converts the rune to byte before writing. Though [bufio.WriteString()](https://github.com/golang/go/blob/master/src/bufio/bufio.go#L703) does not convert the string to bytes directly, it kind of copy bytes from the string before writing.

I do see that proto/buffer.go uses `EncodeXXX()` to write to an in-memory buffer.

Reply:
Yes, there is no right or wrong here, we can go either way. I just followed proto convention here.